marlowe-cli/src/Language/Marlowe/CLI/Transaction.hs

```diff
@@ -982,72 +985,143 @@ buildPublishingImpl
   -> PublishingStrategy era
   -> CoinSelectionStrategy
   -> PrintStats
-  -> m (TxBody era, PublishMarloweScripts MarlowePlutusVersion era)
-buildPublishingImpl connection signingKey expires changeAddress publishingStrategy coinSelectionStrategy (PrintStats printStats) = do
-  pm <- buildPublishScript connection (fromV2TypedValidator marloweValidator) publishingStrategy
-  pp <- buildPublishScript connection (fromV2TypedValidator rolePayoutValidator) publishingStrategy
-
-  let buildPublishScriptTxOut PublishScript{psMinAda, psPublisher, psReferenceValidator} = do
-        referenceScript <- buildReferenceScript $ viScript psReferenceValidator
-        makeTxOut psPublisher Nothing (lovelaceToValue psMinAda) referenceScript
-
-  outputs <-
-    sequence
-      [ buildPublishScriptTxOut pm
-      , buildPublishScriptTxOut pp
-      ]
-
-  (_, inputs, outputs') <-
-    selectCoins
-      connection
-      mempty
-      outputs
-      Nothing
-      changeAddress
-      coinSelectionStrategy
-
-  txBody <-
-    buildBody
-      connection
-      ([] :: [PayFromScript MarlowePlutusVersion])
-      Nothing
-      []
-      inputs
-      outputs'
-      Nothing
-      changeAddress
-      ((0,) <$> expires)
-      [hashSigningKey signingKey]
-      TxMintNone
-      TxMetadataNone
-      printStats
-      False
-
-  let serialiseAddress = T.unpack . C.serialiseAddress
-
-  when printStats $ liftIO do
-    hPutStrLn stderr ""
-    hPutStrLn stderr $
-      "Marlowe script published at address: " <> serialiseAddress (psPublisher pm)
-    hPutStrLn stderr ""
-    hPutStrLn stderr $
-      "Marlowe script hash: "
-        <> show (viHash (psReferenceValidator pm))
-    hPutStrLn stderr ""
-    hPutStrLn stderr $
-      "Marlowe ref script UTxO min ADA: " <> show (psMinAda pm)
-    hPutStrLn stderr ""
-    hPutStrLn stderr $
-      "Payout script published at address: " <> serialiseAddress (psPublisher pp)
-    hPutStrLn stderr ""
-    hPutStrLn stderr $
-      "Payout script hash: "
-        <> show (viHash (psReferenceValidator pp))
-    hPutStrLn stderr ""
-    hPutStrLn stderr $
-      "Payout ref script UTxO min ADA: " <> show (psMinAda pp)
-
-  pure (txBody, PublishMarloweScripts pm pp)
+  -> m ([TxBody era], MarloweScriptsRefs lang era)
+buildPublishingImpl buildupCtx signingKey expires changeAddress publishingStrategy coinSelectionStrategy (PrintStats printStats) = case plutusScriptVersion @lang of
+  PlutusScriptV1 -> throwError "Plutus Script V1 not supported"
+  PlutusScriptV2 -> do
+    let queryCtx = toQueryContext buildupCtx
+    pm <- buildScriptPublishingInfo @lang queryCtx (fromV2TypedValidator marloweValidator) publishingStrategy
+    pp <- buildScriptPublishingInfo @lang queryCtx (fromV2TypedValidator rolePayoutValidator) publishingStrategy
+    po <- buildScriptPublishingInfo @lang queryCtx (fromV2Validator openRoleValidator) publishingStrategy
```

Comment:
So, publishing all three scripts in the same transaction doesn't exceed the transaction-size limit? Even with two, we were pretty close to the limit.

Reply:
@bwbush we are creating here two transactions and handling UTxO set manually - please check the final section of the function Brian.